push/push-reduce-footprint-and-messaging.md

```diff
@@ -0,0 +1,92 @@
+
+## Proposed solution
+
+Refactoring the current UPS into a more modular system, or a system of systems.
+
+### Apache Kafka
+
+Decoupling the process of push delivery by using messaging. Based on the [Google-Summer-of-Code project from 2017](https://github.com/aerogear/aerogear-unifiedpush-server/tree/GSOC_2017_kafka) to replace the internal JMS broker with an external Kafka broker, we can leverage this work and move the different _push sender_ components into their own systems or modules, as individual servers.
+
+One of the big scaling benefits of Apache Kafka is that it's possible to partion a topic, and attach multiple consumers, organzied in consumer groups, to one topic. 
+Any message sent to the topic is processed by exactly a single consumer in each consumer group, subscribed to the topic. The consumers can be deployed on different hosts, for a better horizontal scaling. 
+
+A nice side-effect of using Apache Kafka is that due to its default durability, the system can be used to later process and analyize all traffic, such as all push notifications. Even a replay of the entire message stream on different consumer groups is possible. This is great for event sourcing and also deep analytics of our push payload.
+
+### Sender Modules
+
+Each _Sender component_ would be refactored into a separate java process, that represents a stateless Kafka consumer, subscribed to it's matching topic, which than triggers the push delivery to the 3rd party networks, such as:
+
+```java
+
+consumer.subscribe(Collections.singleton(APNS_TOPIC));
+
+while (isRunning()) {
+  final ConsumerRecords<String, MessageHolderWithTokens> records = consumer.poll(...);
+
+  records.forEach(record -> {
+
+    final MessageHolderWithTokens messageContainer = record.value();
+    sender.sendPushMessage(messageContainer);
+  }
+}
+```
+
+The consumer thread actively polls the push payload (the `MessageHolderWithTokens` object) from the broker, allowing the consumer also to slow down if it can not keep up. Consumers are organized in consumer groups. In Openshift we can simply add more pods of the "apns-push" component, and all consumers (in the different pods) form a specific consumer group, e.g. `apns-consumers`. 
+
+#### Standalone FatJar processes
+
+Since the different sender modules are at the end just stateless Kakfa consumers, they are implemented as standalone java processes, that are subscribed to their matching topic (e.g. APNS_TOPIC, FCM_TOPIC, etc). 
+
+### Smaller Core-UPS
+
+The UPS without JMS can run in the simpler EAP/WildFly configuration. This allows a transformation to WildFly Swarm, using only `CDI`, `JPA` and `JAX-RS` configuration for an smaller server setup, than the vanilla WildFly deployment/standalone model. 
```

Comment:
Could we put CDI, JPA and JAX-RS in the terms above and provide a link or something

Reply:
sure